### DISCUSSION TITLE
Using an index to flatten arrays to a single array.

### DISCUSSION THREAD

**Discussion by dtaalbers** [ravendb/ravendb#17544] [Q&A] [answered]
I have the following setup: https://dotnetfiddle.net/zfZ8Bz

I'd expect the following outcome:

1. The response should be of type result (e.g. matching the Room class)
2. The count should be 5 (5 rooms were added to the two locations).
3. The location properties (name, id) having a value.

I expect this because I have the following code:

```
locations => from location in locations
             from room in location.Rooms
```

But what I am getting is the two locations as a result. (you can also test the index result via the RavenDB studio on http://live-test.ravendb.net/ database Hogwarts). Why is that?

When I add the following as reduce, like this:
```
public class Rooms_Search : AbstractMultiMapIndexCreationTask<Rooms_Search.Result>
{
	public class Result
	{
		public string Id { get; set; }
		public string Name { get; set; }
		public string LocationId { get; set; }
		public string LocationName { get; set; }
		public string Workspace { get; set; }
	}

	public Rooms_Search()
	{
		AddMap<Location>(
			locations => from location in locations
						 from room in location.Rooms
						 select new
						 {
							 Id = room.Id,
							 Name = room.Name,
							 LocationId = location.Id,
							 LocationName = location.Name,
							 Workspace = location.Workspace
						 }
		);

		Reduce = results => results        
			.Select(x => new
			{
				x.Id,
				x.Name,
				x.LocationId,
				x.LocationName,
				x.Workspace
			});

		Index(x => x.Name, FieldIndexing.Search);
	}
}
```
I get the following error :

```
reduce function must contain a group by expression
```

When I change the reduce to this. I get the expected 5 rooms, but the `GroupBy` and `SelectMany` doesn't feel good because I feel like it is unnecessary work on the database part? 
```
Reduce = results => results        
	.GroupBy(x => x.Workspace)
	.SelectMany(x => x)
	.Select(x => new
	{
		x.Id,
		x.Name,
		x.LocationId,
		x.LocationName,
		x.Workspace
	});
```

**Comment by garayx** [comment, marked as answer]
hi

Querying on index is returning the matching document, so the index data is only used for finding the matching documents, if you want to return the actual index data you need to set `Store Fields` (see docs [here](https://ravendb.net/docs/article-page/5.4/csharp/indexes/storing-data-in-index#remarks) and [here](https://ravendb.net/docs/article-page/5.4/csharp/indexes/indexing-linq-extensions#remarks)) in the index. 
Then in the query you need to project the results (https://ravendb.net/docs/article-page/6.0/csharp/client-api/session/querying/how-to-project-query-results#projectinto).

So the index & query should look like this:

Index:
```
    public class Rooms_Search : AbstractMultiMapIndexCreationTask<Rooms_Search.Result>
        {
            public class Result
            {
                public string Id { get; set; }
                public string Name { get; set; }
                public string LocationId { get; set; }
                public string LocationName { get; set; }
                public string Workspace { get; set; }
            }

            public Rooms_Search()
            {
                AddMap<Location>(
                    locations => from location in locations
                        from room in location.Rooms
                        select new
                        {
                            Id = room.Id,
                            Name = room.Name,
                            LocationId = location.Id,
                            LocationName = location.Name,
                            Workspace = location.Workspace
                        }
                );

                Index(x => x.Name, FieldIndexing.Search);
                StoreAllFields(FieldStorage.Yes);
            }
        }
```

Query:

```

                using (var session = store.OpenAsyncSession())
                {
                    // Query data from the index
                    var query = await session.Query<Location, Rooms_Search>()
                        .Where(x => x.Workspace == "hogwarts").ProjectInto<Rooms_Search.Result>()
                        .OrderBy(x => x.Name)
                        .ToListAsync();

                }
```

**Reply by dtaalbers** [reply]
Thank you for the reply!

I tried your suggestion and I get the right amount of records (5 rooms) now. But when projecting it in the `Results` it is using the ID of the location instead of the ID of the room, even though I use the `room.Id` in the Index.

I have updated [my example](https://dotnetfiddle.net/zfZ8Bz). When you run it you see the correct room names in the logs, but you see the wrong IDs. It is using the Location ids. Any idea how to solve that?

I could change some property names etc to fix this, but I would like to understand when I say use `room.Id` for the `Id` in the `Result` class, but then when projecting it uses the `location.Id` for the `Id` in the `Result` class.  What is the point of 
of the following then:
```
select new
{
    Id = room.Id,
    Name = room.Name,
    .... the rest
}

**Comment by garayx** [comment]
The `Id` is kind of a key word and used to pull the document id.

You can open `TrafficWatch` in the studio, send the query, you will see that the query is translated into this RQL:

```
from index 'Rooms/Search' where Workspace = $p0 order by Name select id() as Id, Name, LocationId, LocationName, Workspace
{"p0":"hogwarts"}
```

you can see that `Id` got translated into `id() as Id` which will pull the document id (Location id)

the best option is to use more meaningful name in the index (like `RoomId`) and don't use `Id`:

```
select new
{
    RoomId = room.Id,
    .... the rest
}
``` 

In case you want to continue using Id and pull it from the index data, you will need to use [Session.Advanced.RawQuery](https://ravendb.net/docs/article-page/6.0/csharp/client-api/session/querying/how-to-query#session.advanced.rawquery) and pass the following RQL:

```
from index 'Rooms/Search' where Workspace = "hogwarts" order by Name select Id, Name, LocationId, LocationName, Workspace
```

**Reply by dtaalbers** [reply]
Yeah, understood. I was afraid I had to change the `Id` to `RoomId` indeed. Thank you for the help!
